File: 2019/day02_1.py
# pylint: disable=missing-module-docstring, missing-function-docstring


from typing import Any


def solution(content: str) -> Any:
    program = [int(x) for x in content.split(",")]
    program[1] = 12
    program[2] = 2
    current = 0
    halt = False
    while not halt:
        current, program, halt = execute(current, program)
    return program[0]


def execute(current: int, program: list[int]):
    opcode = program[current]
    if opcode not in [1, 2, 99]:
        raise ValueError(f"Invalid opcode {opcode} at {current} in {program}")

    copy = program.copy()
    halt = False
    if opcode == 1:
        copy[copy[current + 3]] = copy[copy[current + 1]] + copy[copy[current + 2]]
    if opcode == 2:
        copy[copy[current + 3]] = copy[copy[current + 1]] * copy[copy[current + 2]]
    if opcode == 99:
        halt = True

    return (current + 4, copy, halt)


def main() -> None:
    with open("day02.dat", "r", encoding="utf8") as file:
        print(solution(file.read().strip()))


if __name__ == "__main__":
    main()

File: 2019/day02-1.ts
function solution(content: string) {
  return execute(content.split(",").map((item) => parseInt(item)));
}

function execute(program: number[]) {
  let copy = program.slice();
  let halt = false;
  let current = 0;
  copy[1] = 12;
  copy[2] = 2;
  while (!halt) {
    const result = executeOne(current, copy);
    current = result[0];
    copy = result[1];
    halt = result[2];
  }
  return copy[0];
}

function executeOne(
  current: number,
  program: number[]
): [number, number[], boolean] {
  const opcode = program[current];
  const copy = program.slice();
  if (opcode == 1) {
    copy[copy[current + 3]] = copy[copy[current + 1]] + copy[copy[current + 2]];
    return [current + 4, copy, false];
  }
  if (opcode == 2) {
    copy[copy[current + 3]] = copy[copy[current + 1]] * copy[copy[current + 2]];
    return [current + 4, copy, false];
  }
  if (opcode == 99) {
    return [0, copy, true];
  }
  throw Error(
    `Invalid opcode ${opcode} at ${current} in ${program.slice(
      current,
      current + 4
    )}`
  );
}

Deno.readTextFile("day02.dat")
  .then((content) => content.trim())
  .then(solution)
  .then(console.log);
